executor/update.go

```diff
@@ -194,6 +194,12 @@ func (e *UpdateExec) composeNewRow(rowIdx int, oldRow []types.Datum) ([]types.Da
 		if err1 := e.handleErr(assign.Col.ColName, rowIdx, err); err1 != nil {
 			return nil, err1
 		}
+
+		val, err = val.ConvertTo(e.ctx.GetSessionVars().StmtCtx, fields[assign.Col.Index])
```

Comment:
It's better to use https://github.com/pingcap/tidb/blob/a84cce1bcabfdb931a3ebec226d739233645a91b/table/column.go#L149
here.